src/chainstate/coordinator/mod.rs

```diff
@@ -500,20 +540,561 @@ fn dispatcher_announce_burn_ops<T: BlockEventDispatcher>(
     );
 }
 
+fn forget_orphan_stacks_blocks(
+    sort_conn: &DBConn,
+    chainstate_db_tx: &mut DBTx,
+    burn_header: &BurnchainHeaderHash,
+    invalidation_height: u64,
+) {
+    if let Ok(sns) = SortitionDB::get_all_snapshots_for_burn_block(&sort_conn, &burn_header) {
+        for sn in sns.into_iter() {
+            // only retry blocks that are truly in descendant
+            // sortitions.
+            if sn.sortition && sn.block_height > invalidation_height {
+                if let Err(e) = StacksChainState::forget_orphaned_epoch_data(
+                    chainstate_db_tx,
+                    &sn.consensus_hash,
+                    &sn.winning_stacks_block_hash,
+                ) {
+                    warn!(
+                        "Failed to forget that {}/{} is orphaned: {:?}",
+                        &sn.consensus_hash, &sn.winning_stacks_block_hash, &e
+                    );
+                }
+            }
+        }
+    }
+}
+
 impl<'a, T: BlockEventDispatcher, N: CoordinatorNotices, U: RewardSetProvider>
     ChainsCoordinator<'a, T, N, U>
 {
-    pub fn handle_new_stacks_block(&mut self) -> Result<(), Error> {
+    pub fn handle_new_stacks_block(&mut self) -> Result<Option<BlockHeaderHash>, Error> {
         if let Some(pox_anchor) = self.process_ready_blocks()? {
             self.process_new_pox_anchor(pox_anchor)
         } else {
-            Ok(())
+            Ok(None)
+        }
+    }
+
+    /// Get all block snapshots and their PoX IDs at a given burnchain block height.
+    fn get_snapshots_and_pox_ids_at_height(
+        &mut self,
+        height: u64,
+    ) -> Result<Vec<(BlockSnapshot, PoxId)>, Error> {
+        let sort_ids = SortitionDB::get_sortition_ids_at_height(self.sortition_db.conn(), height)?;
+        let ic = self.sortition_db.index_conn();
+
+        let mut ret = Vec::with_capacity(sort_ids.len());
+
+        for sort_id in sort_ids.iter() {
+            let handle = ic.as_handle(sort_id);
+
+            let sn = SortitionDB::get_block_snapshot(&self.sortition_db.conn(), sort_id)?
+                .expect("BUG: have sortition ID without snapshot");
+
+            let pox_id = handle.get_pox_id()?;
+            ret.push((sn, pox_id));
+        }
+
+        Ok(ret)
+    }
+
+    fn handle_affirmation_reorg(&mut self) -> Result<(), Error> {
```

Comment:
this method should have comments, and maybe even be broken into parts.

please annotate errors returned in an `#Errors` section.

Reply:
Added an explainer. Despite its apparent length, the method is already pretty irreducibly compact for what it needs to do.